core/CoreTransaction.ts

```diff
@@ -0,0 +1,117 @@
+import {ConceptManager} from "../api/concept/ConceptManager";
+import {LogicManager} from "../api/logic/LogicManager";
+import {QueryManager} from "../api/query/QueryManager";
+import {ConceptManagerImpl} from "../concept/ConceptManagerImpl";
+import {LogicManagerImpl} from "../logic/LogicManagerImpl";
+import {QueryManagerImpl} from "../query/QueryManagerImpl";
+import {Stream} from "../common/util/Stream";
+import TRANSACTION_CLOSED = ErrorMessage.Client.TRANSACTION_CLOSED;
+
+export class CoreTransaction implements GraknTransaction.Extended {
+    private readonly _session: CoreSession;
+    private readonly _sessionId: string;
+    private readonly _type: TransactionType;
+    private readonly _options: GraknOptions;
+    private _bidirectionalStream: BidirectionalStream;
+    private _conceptManager: ConceptManager;
+    private _logicManager: LogicManager;
+    private _queryManager: QueryManager;
+
+    constructor(session: CoreSession, _sessionId: string, type: TransactionType, options: GraknOptions) {
+        this._session = session;
+        this._sessionId = _sessionId;
+        this._type = type;
+        this._options = options;
+        const rpcClient = this._session.rpc();
+        this._bidirectionalStream = new BidirectionalStream(rpcClient, this._session.requestTransmitter());
+        this._conceptManager = new ConceptManagerImpl(this);
+        this._logicManager = new LogicManagerImpl(this);
+        this._queryManager = new QueryManagerImpl(this);
+    }
+
+    public async open(): Promise<void> {
+        const openReq = RequestBuilder.Transaction.openReq(this._sessionId, this._type.proto(), this._options.proto(), this._session.networkLatency());
+        await this.rpcExecute(openReq, false);
+    }
+
+    public async close(): Promise<void> {
+        await this._bidirectionalStream.close();
+    }
+
+    public async commit(): Promise<void> {
+        const commitReq = RequestBuilder.Transaction.commitReq();
+        try {
+            await this.rpcExecute(commitReq);
+        } finally {
+            await this.close();
+        }
+    }
+
+    public async rollback(): Promise<void> {
+        const rollbackReq = RequestBuilder.Transaction.rollbackReq();
+        await this.rpcExecute(rollbackReq);
+    }
+
+    public concepts(): ConceptManager {
+        return this._conceptManager;
+    }
+
+    public logic(): LogicManager {
+        return this._logicManager;
+    }
+
+    public query(): QueryManager {
+        return this._queryManager;
+    }
+
+    public options(): GraknOptions {
+        return this._options;
+    }
+
+    public type(): TransactionType {
+        return this._type;
+    }
+
+    public isOpen(): boolean {
+        return this._bidirectionalStream.isOpen();
+    }
+
+    public async rpcExecute(request: Transaction.Req, batch?: boolean): Promise<Transaction.Res> {
+        if (!this.isOpen()) throw new GraknClientError(TRANSACTION_CLOSED);
+        const useBatch = batch ? batch : true;
```

Comment:
This is a bug, `useBatch` will ALWAYS be `true`.
```suggestion
        const useBatch = batch !== false;
```